### ISSUE TITLE
FunctionInvocationException when calling a HttpTrigger locally with optional parameters in route

### ISSUE THREAD

**Issue by petmat** [Azure/Azure-Functions#629]
I have a function defined as follows:

```C#
public static class Navigate {
    [FunctionName("Navigate")]
    public static async Task<HttpResponseMessage> Run(
        [HttpTrigger(
            AuthorizationLevel.Function,
            "get",
            Route = "{shortLinkText}/{extraInfo?}")]
            HttpRequestMessage req,
        string shortLinkText,
        string extraInfo = null) {
        // Irrelevant method body omitted
    }
}
```

The extraInfo parameter is defined as an optional parameter but when I run the function locally and then call it `http://localhost:7071/api/5` I get the following error:

```
Microsoft.Azure.WebJobs.Host.FunctionInvocationException : 
Exception while executing function: Navigate ---> System.InvalidOperationException : 
Exception binding parameter 'extraInfo' ---> System.InvalidOperationException : 
Binding data does not contain expected value 'extraInfo'. 
at Microsoft.Azure.WebJobs.Host.Bindings.Data.ClassDataBinding`1.BindAsync(BindingContext context) 
at async Microsoft.Azure.WebJobs.Host.Triggers.TriggeredFunctionBinding`1.BindCoreAsync[TTriggerValue](ValueBindingContext context,Object value,IDictionary`2 parameters) 
End of inner exception at Microsoft.Azure.WebJobs.Host.Executors.DelayedException.Throw() 
at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithWatchersAsync(IFunctionInstance instance,ParameterHelper parameterHelper,TraceWriter traceWriter,CancellationTokenSource functionCancellationTokenSource) 
at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(??) 
at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(??) 
End of inner exception at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() 
at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(??) 
at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.TryExecuteAsync(IFunctionInstance functionInstance,CancellationToken cancellationToken) 
at Microsoft.Azure.WebJobs.Host.Executors.ExceptionDispatchInfoDelayedException.Throw() at async Microsoft.Azure.WebJobs.JobHost.CallAsync(??) 
at async Microsoft.Azure.WebJobs.Script.ScriptHost.CallAsync(String method,Dictionary`2 arguments,CancellationToken cancellationToken) 
at async Microsoft.Azure.WebJobs.Script.WebHost.WebScriptHostManager.HandleRequestAsync(FunctionDescriptor function,HttpRequestMessage request,CancellationToken cancellationToken) 
at async Microsoft.Azure.WebJobs.Script.Host.FunctionRequestInvoker.ProcessRequestAsync(HttpRequestMessage request,CancellationToken cancellationToken,WebScriptHostManager scriptHostManager,WebHookReceiverManager webHookReceiverManager) 
at async Microsoft.Azure.WebJobs.Script.WebHost.Controllers.FunctionsController.<>c__DisplayClass3_0.<ExecuteAsync>b__0(??) 
at async Microsoft.Azure.WebJobs.Extensions.Http.HttpRequestManager.ProcessRequestAsync(HttpRequestMessage request,Func`3 processRequestHandler,CancellationToken cancellationToken) 
at async Microsoft.Azure.WebJobs.Script.WebHost.Controllers.FunctionsController.ExecuteAsync(HttpControllerContext controllerContext,CancellationToken cancellationToken) 
at async System.Web.Http.Dispatcher.HttpControllerDispatcher.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken) 
at async System.Web.Http.Dispatcher.HttpControllerDispatcher.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken) 
at async Microsoft.Azure.WebJobs.Script.WebHost.Handlers.WebScriptHostHandler.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken) 
at async Microsoft.Azure.WebJobs.Script.WebHost.Handlers.SystemTraceHandler.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken) 
at async System.Web.Http.HttpServer.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken)
```

I have:
Azure Functions and Web Jobs Tools 15.0.31201.0
Visual Studio Enterprise 2017 15.5.1

**Comment by kashimiz** [comment]
This didn't repro in the following environment:

- Azure Functions and Web Jobs Tools 15.0.31201.0
- Visual Studio Enterprise 2017 15.5.2

Can you try upgrading your VS2017 install?

**Comment by petmat** [comment]
I'm not sure what happened but I changed tested this again with the 15.5.2 version of Visual Studio 2017 and got the same error. But then I first made the parameter required tested that it worked and after that I tried it as an optional parameter and it still worked! So maybe the problem had something to do with building the project not correctly generating something. I have noticed that I have to often do a rebuild instead of just building the Function App or some things might not work correctly.

But anyways now I got my Function working correctly. I even made another project to test a function with a same route and signature and it worked too so this must have been some temporary glitch. Thanks for the assistance!
